pkg/component/crddeployer/crd.go

```diff
@@ -0,0 +1,155 @@
+// SPDX-FileCopyrightText: SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package crddeployer
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/exp/maps"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/component"
+	"github.com/gardener/gardener/pkg/utils/flow"
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/kubernetes/health"
+	"github.com/gardener/gardener/pkg/utils/retry"
+)
+
+var (
+	// CRDWaitTimeout specifies the total time to wait for CRDs to become ready.
+	CRDWaitTimeout = 15 * time.Second
+)
+
+// crdDeployer is a DeployWaiter that can deploy CRDs and wait for them to be ready.
+type crdDeployer struct {
+	client            client.Client
+	applier           kubernetes.Applier
+	crdNameToManifest map[string]string
+}
+
+// NewCRDDeployer returns a DeployWaiter that can deploy CRDs and wait for them to be ready.
+func NewCRDDeployer(client client.Client, applier kubernetes.Applier, manifests []string) (component.DeployWaiter, error) {
+	crdNameToManifest, err := MakeCRDNameMap(manifests)
+	if err != nil {
+		return nil, err
+	}
+	return &crdDeployer{
+		client:            client,
+		applier:           applier,
+		crdNameToManifest: crdNameToManifest,
+	}, nil
+}
+
+// Deploy deploys the CRDs.
+func (c *crdDeployer) Deploy(ctx context.Context) error {
+	var fns []flow.TaskFn
+
+	for _, resource := range c.crdNameToManifest {
+		fns = append(fns, func(ctx context.Context) error {
+			return c.applier.ApplyManifest(ctx, kubernetes.NewManifestReader([]byte(resource)), kubernetes.DefaultMergeFuncs)
+		})
+	}
+
+	return flow.Parallel(fns...)(ctx)
+}
+
+// Destroy destroys the CRDs.
+func (c *crdDeployer) Destroy(ctx context.Context) error {
+	var fns []flow.TaskFn
+
+	for resourceName, _ := range c.crdNameToManifest {
+		fns = append(fns, func(ctx context.Context) error {
+			crd := &apiextensionsv1.CustomResourceDefinition{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: resourceName,
+				},
+			}
+			return client.IgnoreNotFound(c.client.Delete(ctx, crd))
+		})
+	}
+
+	return flow.Parallel(fns...)(ctx)
+}
+
+// Wait waits for the CRDs to be deployed.
+func (c *crdDeployer) Wait(ctx context.Context) error {
+	return WaitUntilCRDManifestsReady(ctx, c.client, maps.Keys(c.crdNameToManifest))
+}
+
+// WaitCleanup waits for destruction to finish and CRDs to be fully removed.
+func (c *crdDeployer) WaitCleanup(ctx context.Context) error {
+	return WaitUntilCRDManifestsDestroyed(ctx, c.client, maps.Keys(c.crdNameToManifest))
+}
+
+// MakeCRDNameMap returns a map that has the name of the resource as key, and the corresponding manifest as value.
+func MakeCRDNameMap(manifests []string) (map[string]string, error) {
```

Comment:
```suggestion
func generateNameToCRDMap(manifests []string) (map[string]string, error) {
```

Reply:
This function will be used in the future by other packages ([ref](https://github.com/gardener/gardener/pull/10639/files#diff-f509e0b36930e845e5c75c98149f033654d387b64b15ba59982117e82551ad03R49)).

Should I not export it for the time being and export it when the alignment of the other deployers follows?

Reply:
I'm not sure if it's still needed there. Let's not export it for now and discuss again in the next PR.